socks.go

```diff
@@ -54,48 +55,87 @@ const (
 	SOCKS5
 )
 
+//Opt Some optional parameters
+type Opt struct {
+	User        string
+	Password    string
+	DialTimeout time.Duration //default : 5 seconds
+	Timeout     time.Duration //default : 15 seconds
+}
+
 // DialSocksProxy returns the dial function to be used in http.Transport object.
 // Argument socksType should be one of SOCKS4, SOCKS4A and SOCKS5.
 // Argument proxy should be in this format "127.0.0.1:1080".
-func DialSocksProxy(socksType int, proxy string) func(string, string) (net.Conn, error) {
+func DialSocksProxy(socksType int, proxy string, opt *Opt) func(string, string) (net.Conn, error) {
+	if opt == nil {
+		opt = new(Opt)
+	}
+	if opt.Timeout == 0 {
+		opt.Timeout = time.Second * 15
+	}
+	if opt.DialTimeout == 0 {
+		opt.DialTimeout = opt.Timeout
+	}
 	if socksType == SOCKS5 {
 		return func(_, targetAddr string) (conn net.Conn, err error) {
-			return dialSocks5(proxy, targetAddr)
+			return dialSocks5(proxy, targetAddr, opt)
 		}
 	}
 
 	// SOCKS4, SOCKS4A
 	return func(_, targetAddr string) (conn net.Conn, err error) {
-		return dialSocks4(socksType, proxy, targetAddr)
+		return dialSocks4(socksType, proxy, targetAddr, opt)
 	}
 }
 
-func dialSocks5(proxy, targetAddr string) (conn net.Conn, err error) {
+func dialSocks5(proxy, targetAddr string, opt *Opt) (conn net.Conn, err error) {
 	// dial TCP
-	conn, err = net.Dial("tcp", proxy)
+	conn, err = net.DialTimeout("tcp", proxy, opt.DialTimeout)
 	if err != nil {
 		return
 	}
 
 	// version identifier/method selection request
 	req := []byte{
 		5, // version number
-		1, // number of methods
-		0, // method 0: no authentication (only anonymous access supported for now)
+		2, // number of methods
```

Comment:
Since there is no unit test now (my fault), it would be better to keep the header contents exactly like before when the username and password are empty, so that we do not have to worry about anonymous access is broken by the change.